mat64/product.go

```diff
@@ -0,0 +1,190 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import "fmt"
+
+// Product calculates the product of the given factors and places the result in
+// the receiver. The order of multiplication operations is optimized to minimize
+// the number of floating point operations on the basis that all matrix
+// multiplications are general.
+func (m *Dense) Product(factors ...Matrix) {
+	// The operation order optimisation is the naive O(n^3) dynamic
+	// programming approach and does not take into consideration
+	// finer-grained optimisations that might be available.
+	//
+	// TODO(kortschak) Consider using the O(nlogn) or O(mlogn)
+	// algorithms that are available. e.g.
+	//
+	// e.g. http://www.jofcis.com/publishedpapers/2014_10_10_4299_4306.pdf
+	//
+	// In the case that this is replaced, retain this code in
+	// tests to compare against.
+
+	r, c := m.Dims()
+	switch len(factors) {
+	case 0:
+		if r != 0 || c != 0 {
+			panic(ErrShape)
+		}
+		return
+	case 1:
+		m.reuseAs(factors[0].Dims())
+		m.Copy(factors[0])
+		return
+	case 2:
+		// Don't do work that we know the answer to.
+		m.Mul(factors[0], factors[1])
+		return
+	}
+
+	// Check size early, but don't yet allocate data.
+	if !m.isZero() {
+		if fr, _ := factors[0].Dims(); fr != r {
+			panic(ErrShape)
+		}
+		if _, lc := factors[len(factors)-1].Dims(); lc != c {
+			panic(ErrShape)
+		}
+	}
+
+	dims := make([]int, len(factors)+1)
+	dims[0] = r
+	dims[len(dims)-1] = c
+	for i, f := range factors[1:] {
+		_, pc := factors[i].Dims()
+		cr, _ := f.Dims()
+		dims[i+1] = cr
+		if pc != cr {
+			panic(ErrShape)
+		}
+	}
+
+	p := multiplier{
+		factors: factors,
+		dims:    dims,
+		table:   newTable(len(factors)),
+		onStack: make([]bool, len(factors)),
+	}
+	p.optimize()
+	p.multiply()
+
+	m.reuseAs(p.stack[0].Dims())
+	m.Copy(p.stack[0])
+}
+
+// debugProductWalk enables debugging output for Product.
+const debugProductWalk = false
+
+// multiplier performs operation order optimisation and tree traversal.
+type multiplier struct {
+	factors []Matrix
+	dims    []int
+
+	table table
+
+	stack   []*Dense
+	onStack []bool
+}
+
+// optimize determines an optimal matrix multiply operation order.
```

Comment:
Would you mind adding some quick comments?
1) table.at(i,j).cost contains the cost of multiplying the matrices from i to j
2) table.at(i,j).k is which index to subdivide (I think, still parsing.)